server/src-lib/Hasura/GraphQL/Transport/HTTP.hs

```diff
@@ -42,31 +42,27 @@ runGQ reqId userInfo reqHdrs req = do
         E.Tree resolvedPlan unresolvedPlansNE -> do
           let unresolvedPlans = toList unresolvedPlansNE -- it's safe to convert here
           HttpResponse initJson _ <- runLeafPlan resolvedPlan
-          let remoteRels =
-                map
-                  (\(E.QExecPlanUnresolved remoteRelField _) -> remoteRelField)
-                  unresolvedPlans
           let (initValue, remoteBatchInputs) =
-                E.extractRemoteRelArguments
-                  initJson
-                  remoteRels
+                E.extractRemoteRelArguments initJson $
+                  map E.remoteRelField unresolvedPlans
+
+          -- TODO This zip may discard some unresolvedPlans when permissions
+          -- come into play. It's not totally clear to me if this is correct,
+          -- or how it works. Can use Data.Align for safer zips, but it seems like 
+          -- 'extractRemoteRelArguments' should be returning the zipped data.
+          let batchesRemotePlans =
+                -- TODO pass 'G.OperationType' properly when we support mutations, etc.
+                zipWith (E.mkQuery G.OperationTypeQuery) remoteBatchInputs unresolvedPlans 
```

Comment:
If we can get rid of `Batch` (see comment on `data Batch` above), then we won't have to zipWith here.